Named pipelines, | operator and reverse pipeline

jennan opened the pull request:
This PR implements the following changes:
- allow pipelines to be named, to make it easier to sub-select a part of a pipeline after its creation,
- use the | operator to merge pipelines,
- implement a reverse pipeline property to undo the effect of a pipeline.

Few things need to be done to finish this PR:
- add tests
- add a bit of documentation
- add an example notebook (almost done)
-  detect if 2 pipelines with the same name are merged and issue an error

jennan committed:
add name to pipeline and allow retrieval os sub-pipelines using their names

jennan committed:
add a property to reverse the effects of a pipeline

jennan committed:
reverse each step of the pipeline instead of using a class for that

jennan committed:
allow using | operator to combine pipelines

jennan committed:
store pipeline name in steps

Sub-pipelines are now virtually stored inside the pipeline steps names. This
ensure that merging pipelines using | operator works and preserve the names.

jennan committed:
Merge branch 'develop' into named_pipelines

tennlee commented on packages/pipeline/src/pyearthtools/pipeline/controller.py:
```diff
@@ -396,8 +402,28 @@ def steps(
             else:
                 steps_list.append(v)
+        if self.name is not None:
+            for step in steps_list:
+                step.name = self.name
```
'
```suggestion
            for i, step in enumerate(steps_list):
                if not hasattr(step, 'name'):
                    step.name = f'{self.name}.{i}'
```

tennlee replied:
Thoughts: The step could already be a named pipeline, or maybe one day we have named steps.

jennan replied:
For this PR, it was easier to just add a `name` parameter to the class constructor of the Pipeline. Doing so for steps would involve potentially changing all constructors of all Step objects to take an additional `name` parameter... I thought we could look at it later (i.e. future PR) to keep the scope tight for this one.

tennlee replied:
In this case if of the steps is a pipeline with a name, won't your code over-write the name of the pipeline?

jennan replied:
When a pipeline is step, it is not saved inside the new Pipeline object as a step but actually its steps are added to the list of the new Pipeline object being created.

Also, I have changed this part in the end, as it was causing other issues with the | operator and went for explicitely recording named pipelines in a dictionary `self._named` in the Pipeline object.

tennlee replied:
Resolving as this code has now changed

tennlee commented:
I'm sure you've thought about this, but it would be nice to have a tutorial to go with it. I'd suggest one which does some processing in xarray in one named section, that can be used for plotting, then has another named section for the tonumpy steps.

tennlee committed:
Add named pipe demo

tennlee commented:
I took it for a spin. I added a notebook to the branch to show what I tried. The specification of a name and access-by-name seemed to work as expected, but the final joined pipeline didn't work as I expected. Let me know if I've misunderstood how it should be used. The notebook is just at the top-level notebook directory, if it's kept for the final PR, it should be moved into a suitable spot and expanded upon, but I'm also fine it it's just there temporarily to aid in debugging and demonstration.

jennan commented:
> I took it for a spin. I added a notebook to the branch to show what I tried. The specification of a name and access-by-name seemed to work as expected, but the final joined pipeline didn't work as I expected. Let me know if I've misunderstood how it should be used. The notebook is just at the top-level notebook directory, if it's kept for the final PR, it should be moved into a suitable spot and expanded upon, but I'm also fine it it's just there temporarily to aid in debugging and demonstration.

@tennlee Your notebook uses the new features as expected. Can you detail what is not working for `p3`?  it is not obvious for me what would not work and I don't have ERA5LowRes at hand currently.

jennan committed:
add example notebook for the new pipeline features (WIP)

jennan commented:
I have added a notebook explaining the 3 new features (named pipelines, | operator and reversed pipeline). I have put it under `notebooks/pipeline`, and named it `Extras.ipynb` for now, I am happy to change the name of it (but couldn't find a good name when creating it). Also, it looks like the reverse pipeline feature is not working as I thought it was (see last cell), so I have a bit more work to do there too.

tennlee commented:
I'll need to do some more diagnosis I guess. I was able to reproduce your pipeline notebook just fine. With the .reversed, I think it might be better to make it .reversed() if it's returning a new reversed pipeline rather than being the same pipeline with a .reversed attribute. It depends on how you expect people to use it. I'm happy with it staying as an attribute if the ideas is that people will use it that way. Like if someone modifies the state of the pipeline somehow, what happens to the reversed one? e.g. if they insert a new step in the middle, what should the reversed pipeline do? I don't think there's any "best" answer, it just needs to be documented. If .reversed is an attribute which will walk the steps applying undo, that's fine. But if you expect to get a copy of the pipeline with separate state, then it should be a method. I don't have an opinion about which one to choose.

jennan committed:
Revert "store pipeline name in steps"

This reverts commit d8ca63fe45c6c8d943595064ea678355c3c1a47b.

Save nested pipelines in a dictionary.

tennlee commented on packages/pipeline/src/pyearthtools/pipeline/controller.py:
```diff
@@ -182,6 +182,10 @@ def _get_tree(
         return graph, prior_step
 
 
+class NotReversiblePipeline(RuntimeError):
```
Maybe call this "ForwardOnly"? Just a thought

jennan committed:
Revert "reverse each step of the pipeline instead of using a class for that"

This reverts commit d075868afd2eaed9d90948c42caa6f9efc603f45.

Trying to reverse a pipeline by reversing each underlying operation is
more complicated than expected, due to operations that loose their state
when transposed using `.T`.  One example is the ToNumpy operation, which
has multiple private members that won't be initialised after transposition,
leading to a non-functional reversed operation.  As multiple operations might
have the save behavior, it seems easier to reverse the whole pipeline as a
black box instead of each step.

jennan committed:
implement pipeline reverse as an Operation instead of a Pipeline

Treat the ReversePipeline class as an opaque class, not meant to expose
steps of of the underlying pipeline, etc. The Operation class provides the
right API to capture the minimum operations this class needs (apply and undo)
and makes it easy to integrate the ReversedPipeline in an existing pipeline.

jennan commented:
@tennlee I think the PR is almost ready. Few things that we could change before merging:

- Do you want to keep the NamedPipe.ipynb example notebook in addition/instead of the notebook Extras.ipynb?
- `pipeline.reversed` is a bit less safe now, as I am not checking that every step is an Operation with an working `.T` attribute. Would it make sense to add some additional exceptions if/when people use a non-reversible pipeline?

jennan committed:
remove test notebook, renamed example notebook

jennan committed:
add example of end to end pipeline, explain undo logic of pipelines

jennan commented:
So, it turned out that the `Pipeline.undo` method does a bit more than I thought, removing the need (and possibility) for a thorough inspection of the pipeline object in the `ReversedPipeline` class constructor. Indeed, `Pipeline.undo` will automatically skip the steps that are transforms for example. Additionally, steps themselves might have an undo method that does nothing, this is the case of the `TemporalRetrieval` step for example.

All in all, it means that a preprocessing pipeline (including data sources) can be reversed without the need to carefully split and reassemble the reversible steps. I have updated the example notebook to show this at the end with a simple persistence-like model.

I think this also means that users would need a good understanding of the undo behavior of each steps to avoid bad surprises.

jennan committed:
Merge branch 'develop' into named_pipelines

jennan committed:
fix example notebook for new CoordinateFlatten API

CoordinateFlatten only supports sequence of strings as input parameters now.

jennan committed:
add ReversedPipeline in the API docs

jennan committed:
add patterns notebook to the gallery

jennan committed:
remove extra empty cell

jennan commented:
The API and notebook docs have been updated, @tennlee this PR is now ready :).
